pkg/operator/operator.go

```diff
@@ -143,42 +145,44 @@ func (c *vSphereProblemDetectorController) runChecks(ctx context.Context) (time.
 		return 0, err
 	}
 
-	var results []checkResult
 	checkContext := &check.CheckContext{
 		Context:    ctx,
 		VMConfig:   vmConfig,
 		VMClient:   vmClient,
 		KubeClient: c,
 	}
 
+	checkRunner := NewCheckThreadPool(parallelVSPhereCalls)
+	resultCollector := NewResultsCollector()
 	var errs []error
-	clusterResults, err := c.runClusterChecks(checkContext)
-	if err != nil {
+	if err := c.addClusterChecks(checkContext, checkRunner, resultCollector); err != nil {
 		errs = append(errs, err)
 	}
-	if clusterResults != nil {
-		results = append(results, clusterResults...)
-	}
-
-	nodeResults, err := c.runNodeChecks(checkContext)
-	if err != nil {
+	if err := c.addNodeChecks(checkContext, checkRunner, resultCollector); err != nil {
 		errs = append(errs, err)
 	}
-	if nodeResults != nil {
-		results = append(results, nodeResults...)
+
+	klog.V(4).Infof("Waiting for all checks")
+	if err := checkRunner.Wait(ctx); err != nil {
+		return 0, err
 	}
 
+	klog.V(4).Infof("All checks complete")
+
+	results, resultErrors := resultCollector.Collect()
 	c.reportResults(results)
 	c.lastResults = results
 	c.lastCheck = time.Now()
 	var nextDelay time.Duration
-	finalErr := errors.NewAggregate(errs)
+	// Join errors from add*Check and all result errors to determine the next check delay
+	finalErr := errors.NewAggregate(append(errs, resultErrors...))
 	if finalErr != nil {
```

Comment:
as far as I could see - `finalErr` is entirely used as a boolean. Do we really need to collect all these errors in a big array then?

Reply:
Ok, reworked `Collect` to return a boolean